src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IsPatternOperator.cs

```diff
@@ -108,29 +108,27 @@ public new void Free()
 
             internal BoundExpression LowerGeneralIsPattern(BoundIsPatternExpression node, BoundDecisionDag decisionDag)
             {
+                Debug.Assert(node.Type is { SpecialType: SpecialType.System_Boolean });
+
                 _factory.Syntax = node.Syntax;
-                var resultBuilder = ArrayBuilder<BoundStatement>.GetInstance();
+                var sideEffectsBuilder = ArrayBuilder<BoundExpression>.GetInstance();
                 var inputExpression = _localRewriter.VisitExpression(node.Expression);
-                decisionDag = ShareTempsIfPossibleAndEvaluateInput(decisionDag, inputExpression, resultBuilder, out _);
+                decisionDag = ShareTempsAndEvaluateInput(inputExpression, decisionDag, sideEffectsBuilder.Add, out _);
 
                 // lower the decision dag.
+                var resultBuilder = ArrayBuilder<BoundStatement>.GetInstance();
                 ImmutableArray<BoundStatement> loweredDag = LowerDecisionDagCore(decisionDag);
-                resultBuilder.Add(_factory.Block(loweredDag));
-                Debug.Assert(node.Type is { SpecialType: SpecialType.System_Boolean });
-                LocalSymbol resultTemp = _factory.SynthesizedLocal(node.Type, node.Syntax, kind: SynthesizedLocalKind.LoweringTemp);
-                LabelSymbol afterIsPatternExpression = _factory.GenerateLabel("afterIsPatternExpression");
-                LabelSymbol trueLabel = node.WhenTrueLabel;
-                LabelSymbol falseLabel = node.WhenFalseLabel;
-                if (_statements.Count != 0)
-                    resultBuilder.Add(_factory.Block(_statements.ToArray()));
-                resultBuilder.Add(_factory.Label(trueLabel));
-                resultBuilder.Add(_factory.Assignment(_factory.Local(resultTemp), _factory.Literal(true)));
-                resultBuilder.Add(_factory.Goto(afterIsPatternExpression));
-                resultBuilder.Add(_factory.Label(falseLabel));
-                resultBuilder.Add(_factory.Assignment(_factory.Local(resultTemp), _factory.Literal(false)));
-                resultBuilder.Add(_factory.Label(afterIsPatternExpression));
-                _localRewriter._needsSpilling = true;
-                return _factory.SpillSequence(_tempAllocator.AllTemps().Add(resultTemp), resultBuilder.ToImmutableAndFree(), _factory.Local(resultTemp));
+                resultBuilder.AddRange(loweredDag);
+                resultBuilder.AddRange(_statements);
+                return _factory.Sequence(
+                    _tempAllocator.AllTemps(),
+                    sideEffectsBuilder.ToImmutableAndFree(),
```

Comment:
>sideEffectsBuilder.ToImmutableAndFree() [](http://example.com/codeflow?start=20&length=39)

If I understand correctly, leaving these side-effects outside of ```BoundLoweredIsPatternExpression``` guarantees that nothing inside it is going to trigger a spilling. If that is the case, I think it is worth adding a comment regarding that. Also, it would be good to add a DEBUG validation in ```SpillSequenceSpiller``` to catch scenarios that violate the assumption.